src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs

```diff
@@ -0,0 +1,316 @@
+            }
+
+            _minimumBufferSize = options.MinimumBufferSize;
+            _pool = options.Pool == MemoryPool<byte>.Shared ? null : options.Pool;
+            _bufferSegmentPool = new BufferSegmentStack(InitialSegmentPoolSize);
+        }
+
+        /// <summary>
+        /// Gets the inner stream that is being written to.
+        /// </summary>
+        public Stream InnerStream { get; }
+
+        /// <inheritdoc />
+        public override void Advance(int bytes)
+        {
+            if ((uint)bytes > (uint)_tailMemory.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.bytes);
+            }
+
+            _tailBytesBuffered += bytes;
+            _bytesBuffered += bytes;
+            _tailMemory = _tailMemory.Slice(bytes);
+        }
+
+        /// <inheritdoc />
+        public override Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            if (_isCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (sizeHint < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.sizeHint);
+            }
+
+            AllocateMemory(sizeHint);
+
+            return _tailMemory;
+        }
+
+        /// <inheritdoc />
+        public override Span<byte> GetSpan(int sizeHint = 0)
+        {
+            if (_isCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (sizeHint < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.sizeHint);
+            }
+
+            AllocateMemory(sizeHint);
+
+            return _tailMemory.Span;
+        }
+
+        private void AllocateMemory(int sizeHint)
+        {
+            if (_head == null)
+            {
+                // We need to allocate memory to write since nobody has written before
+                BufferSegment newSegment = AllocateSegment(sizeHint);
+
+                // Set all the pointers
+                _head = _tail = newSegment;
+                _tailBytesBuffered = 0;
+            }
+            else
+            {
+                int bytesLeftInBuffer = _tailMemory.Length;
+
+                if (bytesLeftInBuffer == 0 || bytesLeftInBuffer < sizeHint)
+                {
+                    if (_tailBytesBuffered > 0)
+                    {
+                        // Flush buffered data to the segment
+                        _tail.End += _tailBytesBuffered;
+                        _tailBytesBuffered = 0;
+                    }
+
+                    BufferSegment newSegment = AllocateSegment(sizeHint);
+
+                    _tail.SetNext(newSegment);
+                    _tail = newSegment;
+                }
+            }
+        }
+
+        private BufferSegment AllocateSegment(int sizeHint)
```

Comment:
Move this logic to BufferSegment to avoid duplication?

Reply:
That's tricky since the logic for creating an new list from empty (head == null) is different in the Pipe vs this impl. This has 2 pointers (head and tail) and the Pipe has 3 pointers (readHead, readTail and writeHead)